libnopegl/src/pipeline_compat.c

```diff
@@ -401,6 +401,58 @@ int ngli_pipeline_compat_update_buffer(struct pipeline_compat *s, int32_t index,
     return 0;
 }
 
+static int select_next_available_bindgroup(struct pipeline_compat *s)
+{
+    struct gpu_ctx *gpu_ctx = s->gpu_ctx;
+
+    s->cur_bindgroup = s->bindgroups[s->next_bindgroup++];
+    s->next_bindgroup %= s->nb_bindgroups;
+
+    /* If the selected bindgroup is not in used by any command buffer, select it */
+    if (s->cur_bindgroup->rc.count == 1)
+        return 0;
+
+    /* Otherwise, increase our bindgroup pool */
+    size_t nb = s->nb_bindgroups;
+    size_t nb_bindgroups = s->nb_bindgroups * 2;
+    struct bindgroup **bindgroups = ngli_realloc(s->bindgroups, nb_bindgroups, sizeof(*bindgroups));
+    if (!bindgroups)
+        return NGL_ERROR_MEMORY;
+
+    memset(bindgroups + s->nb_bindgroups, 0, s->nb_bindgroups * sizeof(*bindgroups));
+
+    s->bindgroups = bindgroups;
+    s->nb_bindgroups = nb_bindgroups;
+    s->next_bindgroup = nb;
+
+    for (size_t i = nb; i < s->nb_bindgroups; i++) {
+        s->bindgroups[i] = ngli_bindgroup_create(gpu_ctx);
+        if (!s->bindgroups[i])
+            return NGL_ERROR_MEMORY;
+
+        struct bindgroup_params params = {
+            .layout      = s->bindgroup_layout,
+            .textures    = s->textures,
+            .nb_textures = s->nb_textures,
+            .buffers     = s->buffers,
+            .nb_buffers  = s->nb_buffers,
+        };
+
+        int ret = ngli_bindgroup_init(s->bindgroups[i], &params);
+        if (ret < 0)
+            return ret;
+    }
+
+    /* And select the first bindgroup from our newly allocated bindgroups */
+    s->cur_bindgroup = s->bindgroups[s->next_bindgroup++];
+    s->next_bindgroup %= s->nb_bindgroups;
```

Comment:
A bit weird to have this done twice in this codepath (first occurrence at the beginning)

Reply:
I tried to split the logic in another function but it does not improve the overall readability and adds an indirection.

Reply:
Changed to darray and simplified the logic.